README.rst
.. caution::

    This repository is no longer maintained and thus it got archived.

    If you want to work on it please open a ticket in
    https://github.com/zopefoundation/meta/issues requesting its unarchival.

The purpose of this package is to define, populate and use multiple IComponents
instances using filesystem-based development in other words, Python code and
ZCML.


CHANGES.rst
=========
 CHANGES
=========

3.1 (unreleased)
================

- Nothing changed yet.


3.0 (2023-02-09)
================

- Drop support for Python 2.7, 3.4, 3.5, 3.6.

- Add support for Python 3.8, 3.9, 3.10, 3.11.

- Make tests compatible with ``zope.component >= 5``.


2.2.0 (2018-10-19)
==================

- Add support for Python 3.7.

- Drop support for Python 3.3.


2.1.0 (2017-05-03)
==================

- Add support for Python 3.4, 3.5, 3.6 and PyPy.

- Remove test dependency on ``zope.app.testing`` and
  ``zope.app.zcmlfiles``, among others.


2.0.0 (2012-11-17)
==================

- zope.configuration changed action tuples to action dicts. This version works
  with the new action dict given from zope.configuration since version 3.8.0.
  This version is not compatible with zope.configuration version less then
  3.8.0


1.3.0 (2010-10-28)
==================

- Fundamental change in the way how baseregistry hooks into ZCA.
  Now it uses hooks.setSite, which requires that zope.component hooks
  are in place. Usually they are installed by zope.app.appsetup.
  Unless you use zope.app.appsetup, install the hooks with
  zope.component.hooks.setHooks().
  This applies to zope.component versions >= 3.9.4.


1.2.0 (2009-12-27)
==================

- Moved browser dependencies to zmi extras


1.1.0 (2009-03-19)
==================

- Fix base registry management form failure in case, when a site has its
  parent's local site manager (that isn't registered as utility) in its
  __bases__.

- Use zope.site instead of zope.app.component.

- Drop unused dependencies on zope.app.i18n and zope.app.pagetemplate.


1.0.0 (2008-01-24)
==================

- Initial Release


src/z3c/baseregistry/browser/tests.py
##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Test setup for Base setting UI tests.

"""

__docformat__ = "reStructuredText"
import doctest
import unittest

import zope.component
import zope.interface
from zope.app.wsgi.testlayer import BrowserLayer
from zope.browsermenu.menu import getFirstMenuItem
from zope.interface import implementer
from zope.publisher.browser import BrowserView
from zope.publisher.interfaces.browser import IBrowserPublisher
from zope.testbrowser.wsgi import TestBrowserLayer
from zope.testing import renormalizing

from z3c.baseregistry import baseregistry
from z3c.baseregistry import browser as z3c_browser


custom = baseregistry.BaseComponents(
    zope.component.globalSiteManager, 'custom')


class IExample(zope.interface.Interface):
    name = zope.interface.Attribute('Name of Example')


@zope.interface.implementer(IExample)
class Example:

    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return '<{} {!r}>'.format(self.__class__.__name__, self.name)


example1 = Example('example1')
example2 = Example('example2')
example3 = Example('example3')
example4 = Example('example4')


def addBasesSelection(browser, bases):
    # Get the form
    from webtest.forms import MultipleSelect
    from zope.testbrowser.browser import ListControl

    form = browser.getForm('zc.page.browser_form')
    webtest_form = form._form  # XXX: Private API

    # Create the select tag
    webtest_select = MultipleSelect(webtest_form, None,
                                    name="form.__bases__",
                                    pos=0,
                                    id="form.__bases__")
    # Add the options.
    # Be careful to keep the option indexes in the order that matches
    # the test.
    options = [(x, False, x) for x in bases]
    webtest_select.options.extend(options)
    # Select them.
    webtest_select.select_multiple(texts=bases)

    # Add the select tag to the form
    webtest_form.fields['form.__bases__'] = [webtest_select]
    webtest_form.field_order.append(('form.__bases__', webtest_select))
    # And the browser
    select = ListControl(webtest_select, form, 'select', browser)
    form.controls.append(select)


@implementer(IBrowserPublisher)
class ManagementViewSelector(BrowserView):
    """View that selects the first available management view.

    Support 'zmi_views' actions like: 'javascript:alert("hello")',
    '../view_on_parent.html' or '++rollover++'.
    """
    # Copied from zope.app.publication
    # Simplified to assert just the test case we expect.

    def browserDefault(self, request):
        return self, ()

    def __call__(self):
        item = getFirstMenuItem('zmi_views', self.context, self.request)
        assert item
        redirect_url = item['action']
        if not redirect_url.lower().startswith(('../', 'javascript:', '++')):
            self.request.response.redirect(redirect_url)
            return ''
        raise AssertionError("Should not get here")  # pragma: no cover


class LoginLogout:
    # Dummy implementation of zope.app.security.browser.auth.LoginLogout

    def __call__(self):
        return None


class _Z3CRegistryLayer(TestBrowserLayer,
                        BrowserLayer):
    pass


Z3CRegistryLayer = _Z3CRegistryLayer(z3c_browser)


def test_suite():

    readme = doctest.DocFileSuite(
        'README.rst',
        globs={
            'IExample': IExample,
            'addBasesSelection': addBasesSelection,
            'getRootFolder': Z3CRegistryLayer.getRootFolder,
        },
        optionflags=doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE,
        checker=renormalizing.RENormalizing(),
    )
    readme.layer = Z3CRegistryLayer
    suite = unittest.TestSuite((readme,))
    return suite


if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


src/z3c/baseregistry/tests/testdoc.py
##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Base Components test setup."""
import doctest
import unittest

from zope.testing import module
from zope.testing import renormalizing
from zope.testing.cleanup import CleanUp


def setUp(test):
    CleanUp().setUp()
    module.setUp(test, name='README')


def tearDown(test):
    module.tearDown(test, name='README')
    CleanUp().tearDown()


def test_suite():
    return unittest.TestSuite((
        doctest.DocFileSuite(
            '../README.rst',
            setUp=setUp, tearDown=tearDown,
            optionflags=doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS,
            checker=renormalizing.RENormalizing(),
        ),
    ))


src/z3c/baseregistry/tests/test_zcml.py
##############################################################################
#
# Copyright (c) 2017 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################

import unittest

from z3c.baseregistry import zcml


class TestActionsProxy(unittest.TestCase):

    def _makeOne(self, list_=None):
        list_ = list_ or []
        return list_, zcml.ActionsProxy(list_, self)

    def test_set_decorates(self):
        list_, proxy = self._makeOne(['abc'])
        proxy[0] = {'discriminator': 'foo'}
        self.assertEqual([{'discriminator': (self, 'foo')}], list_)

    def test_slice_decorates(self):
        list_, proxy = self._makeOne(['abc'])
        proxy[:] = [{'discriminator': 'foo'}]
        self.assertEqual([{'discriminator': (self, 'foo')}], list_)

    def test_iadd_decorates(self):
        list_, proxy = self._makeOne()
        proxy += [{'discriminator': 'foo'}]
        self.assertEqual([{'discriminator': (self, 'foo')}], list_)

    def test_insert_decorates(self):
        list_, proxy = self._makeOne(['abc'])
        proxy.insert(0, {'discriminator': 'foo'})
        self.assertEqual([{'discriminator': (self, 'foo')}, 'abc'], list_)

    def test_extend_decotares(self):
        list_, proxy = self._makeOne(['abc'])
        proxy.extend(({'discriminator': 'foo'},))
        self.assertEqual(['abc', {'discriminator': (self, 'foo')}], list_)

    def test_proxy(self):
        list_, proxy = self._makeOne()
        self.assertEqual(list_.index, proxy.index)

    def test_len(self):
        list_, proxy = self._makeOne(['abc'])
        self.assertEqual(len(list_), len(proxy))
        self.assertEqual(1, len(proxy))


def test_suite():
    return unittest.defaultTestLoader.loadTestsFromName(__name__)


src/z3c/baseregistry/baseregistry.py
##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Base Components support

"""
__docformat__ = "reStructuredText"
from zope.component import globalregistry
from zope.interface.interfaces import IComponents


def BC(components, name):
    return components.getUtility(IComponents, name)


class BaseComponents(globalregistry.BaseGlobalComponents):
    """An ``IComponents`` implementation that serves as base for other
    components."""

    def __init__(self, parent, *args, **kw):
        self.__parent__ = parent
        super().__init__(*args, **kw)

    def __reduce__(self):
        # Global site managers are pickled as global objects
        return BC, (self.__parent__, self.__name__)


src/z3c/baseregistry/browser/base.py
##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Setting bases for local sites.

"""

__docformat__ = "reStructuredText"
import zope.interface
from zope import component
from zope.component import globalregistry
from zope.formlib import form
from zope.i18nmessageid import ZopeMessageFactory as _
from zope.interface.interfaces import IComponents
from zope.schema.vocabulary import SimpleTerm
from zope.schema.vocabulary import SimpleVocabulary
from zope.security.proxy import removeSecurityProxy
from zope.site.interfaces import ILocalSiteManager


BASENAME = _('-- Global Base Registry --')
PARENTNAME = _('-- Parent Local Registry --')


@zope.interface.provider(zope.schema.interfaces.IVocabularyFactory)
class BaseComponentsVocabulary(SimpleVocabulary):
    """A vocabulary for ``IComponents`` utilities."""

    def __init__(self, context):
        terms = []
        utils = set()

        # add available registry utilities
        for name, util in component.getUtilitiesFor(IComponents, context):
            terms.append(SimpleTerm(util, name))
            utils.add(util)

        # add location parent registry if any
        lsm = [removeSecurityProxy(sm)
               for sm in context.__bases__
               if sm not in utils and ILocalSiteManager.providedBy(sm)]
        lsm = [SimpleTerm(x, PARENTNAME) for x in lsm[:1]]
        terms.extend(lsm)

        # add the base registry
        terms.append(SimpleTerm(globalregistry.base, BASENAME))

        super().__init__(terms)


class IComponentsBases(zope.interface.Interface):
    """An interface describing the bases API of the IComponents object."""

    __bases__ = zope.schema.List(
        title=_('Bases'),
        description=_('The base components registires of this registry.'),
        value_type=zope.schema.Choice(vocabulary='Base Components'),
        required=True)


class SetBasesPage(form.EditForm):
    """A page to set the bases of a local site manager"""
    form_fields = form.FormFields(IComponentsBases)


src/z3c/baseregistry/zcml.py
##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""ZCML Implementation to populate base registries.

"""
__docformat__ = "reStructuredText"
import zope.component.globalregistry
import zope.component.hooks
import zope.configuration.config
import zope.configuration.fields
import zope.interface
from zope.configuration.exceptions import ConfigurationError


class IRegisterInDirective(zope.interface.Interface):
    """Use the specified registry for registering the contained components."""

    registry = zope.configuration.fields.GlobalObject(
        title="Registry",
        description="Python path to the registry to use.",
        required=True)


class ActionsProxy:
    """A proxy object for the actions list to decorate the incoming actions."""

    original = None
    registry = None

    def __init__(self, original, registry):
        self.original = original
        self.registry = registry

    def __decorate(self, action):
        # handle action dict
        # (was a tuple before 2.0, see zope.configuration 3.8 for changes)
        # discriminator is a tuple like:
        # ('utility',
        #  <InterfaceClass zope.component.interfaces.IFactory>,
        #  'my.package.interfaces.IMyInterface')
        # in the sample above this means we need to prepend our registry
        # to the existing discriminator.
        discriminator = action.get('discriminator', None)
        if discriminator is not None:
            # replace the first part from the existing descriminator tuple
            # with our registry
            action['discriminator'] = (self.registry, discriminator)
        return action

    def __setitem__(self, i, item):
        if isinstance(i, slice):
            item = [self.__decorate(x) for x in item]
        else:
            item = self.__decorate(item)
        self.original.__setitem__(i, item)

    def __iadd__(self, other):
        other = [self.__decorate(item) for item in other]
        self.original.__iadd__(other)

    def append(self, item):
        self.original.append(self.__decorate(item))

    def insert(self, i, item):
        self.original.insert(i, self.__decorate(item))

    def extend(self, other):
        other = [self.__decorate(item) for item in other]
        self.original.extend(other)

    def __len__(self):
        return len(self.original)

    def __getattr__(self, name):
        return getattr(self.original, name)


class FakeBaseRegistrySite:
    """This a minimal fake Site, the only responsibility it has
    is to store our registry as a SiteManager and return it later.
    This is needed to fool siteinfo via setSite, zope.component.zcml.handler
    will grab the registry via zope.component.getSiteManager() then."""

    def __init__(self, sm):
        self.sm = sm

    def getSiteManager(self):
        return self.sm


def setActiveRegistry(context, registry):
    context.original = zope.component.hooks.getSite()
    fakeSite = FakeBaseRegistrySite(registry)
    zope.component.hooks.setSite(fakeSite)


def resetOriginalRegistry(context):
    zope.component.hooks.setSite(context.original)


class RegisterIn(zope.configuration.config.GroupingContextDecorator):

    # Marker that this directive has been used in the path
    registryChanged = True

    # Storage for the original site
    original = None

    def __init__(self, context, registry, **kw):
        if hasattr(context, 'registryChanged') and context.registryChanged:
            raise ConfigurationError(
                'Nested ``registerIn`` directives are not permitted.')

        super().__init__(context, **kw)
        self.registry = registry
        self.actions = ActionsProxy(context.actions, registry)

    def before(self):
        self.context.action(
            discriminator=None,
            callable=setActiveRegistry,
            args=(self, self.registry)
        )

    def after(self):
        self.context.action(
            discriminator=None,
            callable=resetOriginalRegistry,
            args=(self,)
        )
